src/main/scala/chisel3/util/experimental/BoringUtils.scala

```diff
@@ -193,4 +204,88 @@ object BoringUtils {
     sinks.foreach(addSink(_, genName, true, true))
     genName
   }
+
+  /** Access a source [[Data]] that may or may not be in the current module.  If
+    * this is in a child module, then create ports to allow access the
+    * requrested source.
+    */
+  def bore(source: Data)(implicit si: SourceInfo): Data = {
+    import reflect.DataMirror
+    def parent(d: Data): experimental.BaseModule = d.topBinding.location.get
+    def purePortType = if (DataMirror.hasOuterFlip(source)) Flipped(chiselTypeOf(source)) else chiselTypeOf(source)
+    def emptyWireOnError = Wire(purePortType) //return empty wire to continue elaboration to collect other errors
+
+    experimental.requireIsHardware(source)
+    val thisModule = Builder.currentModule.get
+    source.topBindingOpt match {
+      case None =>
+        Builder.error(s"Cannot bore from ${source._errorContext}")
+        return emptyWireOnError
+      case Some(internal.CrossModuleBinding) =>
+        Builder.error(
+          s"Cannot bore across a Definition/Instance boundary:${thisModule._errorContext} cannot access ${source}"
+        )
+        return emptyWireOnError
+      case _ => // Actually bore
+    }
+    if (parent(source) == thisModule) {
+
+      /** No boring to do */
+      return source
+    }
+
+    val lca = DataMirror.leastCommonAncestorModule(source, thisModule)
+    if (lca.isEmpty) {
+      Builder.error(s"Cannot bore from $source to ${thisModule.name}, as they do not share a least common ancestor")
+      return Wire(purePortType) //return empty wire to continue elaboration to collect other errors
+    }
+
+    val upPath = DataMirror.modulePath(source, lca)
+    val downPath = DataMirror.modulePath(thisModule, lca)
+    //println(lca)
+    //println(upPath)
+    //println(downPath)
+    val lcaSource = upPath.foldLeft(source) {
+      case (rhs, module) =>
+        chisel3.experimental.skipPrefix { // so `lcaSource` isn't in the name of the secret port
+          if (module.toDefinitionCalled) {
+            Builder.error(
+              s"Can only bore into modules that have not been converted to a Definition: ${module._errorContext}"
+            )
+            return emptyWireOnError
+          }
+
+          /** create a port. */
+          val bore = SecretIO(purePortType, module)
+          module.addSecretIo(bore)
+          module.asInstanceOf[RawModule].secretConnection(bore, rhs)
+          bore
+        }
+    }
+    val sink = if (downPath.isEmpty) { lcaSource }
+    else {
+      downPath.reverse.foldLeft(lcaSource) {
+        case (rhs, module) =>
+          chisel3.experimental.skipPrefix { // so `lcaSource` isn't in the name of the secret port
+            if (module.toDefinitionCalled) {
+              Builder.error(
+                s"Can only bore into modules that have not been converted to a Definition: ${module._errorContext}"
+              )
```

Comment:
Improving this to track what line did the `toDefinition` would help with debugability. That would make a nice separate PR!

Reply:
I'll add it to this PR, it's not too hard.